riscv/mmu.cc

```diff
@@ -149,6 +149,23 @@ void mmu_t::load_slow_path(reg_t addr, reg_t len, uint8_t* bytes, uint32_t xlate
       throw *matched_trigger;
   }
 
+  if (unlikely(addr & (len-1))) {
+    if (require_alignment) {
+      load_reserved_address_misaligned(addr);
+    } else {
+      reg_t value = misaligned_load(addr, len, xlate_flags);
+      for (size_t i = 0; i < len; i++) {
+        if (target_big_endian) {
+          bytes[len - i - 1] = value & 0xff;
+        } else {
+          bytes[i] = value & 0xff;
+        }
+        value >>= 8;
+      }
```

Comment:
Can't this just be `memcpy(bytes, &value, len)`?

Reply:
I think that would do the wrong thing if the host endianness differs from the endianness of the simulated target.

Reply:
It seems to me the problem is that this approach handles endianness 3 times (within `misaligned_load`, right here, and upon return from this function).  While correct, it's quite confusing.  We can get rid of two of them by replacing lines 156 - 164 with

    for (size_t i = 0; i < len; i++)
      bytes[i] = load_uint8(addr + i);

and then the endian swap will occur on return from this function.

Reply:
Hmm... but then you won't get the exception if misaligned handling is disabled. You'd end up copying misaligned_load() into load_slow_path(), and then you have the same problem that @scottj97 pointed out about duplicating code which leads to problems down the road when one copy is changed and not another.

I agree this is a mess. I had started to change load_slow_path() to be a template and return a type, like load_fast() does. Then only load_fast() needs to worry about endianness, and the code gets a bit more consistent as well. But then I discarded it as too big a change that is unrelated.

Reply:
You might’ve missed that I edited my post. I added back the call to misaligned_load, which we evaluate only for the side effects. It’s weird, but it’s way less of a mess.

Reply:
Ah. That works. Implemented.